src/core/processing-queue.js

```diff
@@ -21,69 +20,91 @@ import {
 
 let priorityCount = 0;
 let unitSampler;
+const taskQueue = [];
 
 /**
- * Advances the ProcessingQueue to the next item if it is ready.
- * @param {Boolean} last
+ * Advances the taskQueue to the next task. If the taskQueue is empty,
+ * process the testQueue
  */
 function advance() {
+	advanceTaskQueue();
+
+	if ( !taskQueue.length ) {
+		advanceTestQueue();
+	}
+}
+
+/**
+ * Advances the taskQueue to the next task if it is ready and not empty.
+ */
+function advanceTaskQueue() {
 	const start = now();
 	config.depth = ( config.depth || 0 ) + 1;
 
-	while ( config.queue.length && !config.blocking ) {
+	while ( taskQueue.length && !config.blocking ) {
 		const elapsedTime = now() - start;
 
 		if ( !defined.setTimeout || config.updateRate <= 0 || elapsedTime < config.updateRate ) {
-			if ( priorityCount > 0 ) {
-				priorityCount--;
-			}
-
-			config.queue.shift()();
+			const task = taskQueue.shift();
+			task();
 		} else {
 			setTimeout( advance );
 			break;
 		}
 	}
 
 	config.depth--;
+}
 
+/**
+ * Advance the testQueue to the next test to process. Call done() if testQueue completes.
+ */
+function advanceTestQueue() {
 	if ( !config.blocking && !config.queue.length && config.depth === 0 ) {
 		done();
+		return;
 	}
-}
 
-function addToQueueImmediate( callback ) {
-	if ( objectType( callback ) === "array" ) {
-		while ( callback.length ) {
-			addToQueueImmediate( callback.pop() );
-		}
+	const testTasks = config.queue.shift();
+	addToTaskQueue( testTasks() );
+	advance();
+}
 
-		return;
-	}
+/**
+ * Enqueue the tasks for a test into the task queue.
+ * @param {Array} tasksArray
+ */
+function addToTaskQueue( tasksArray ) {
+	taskQueue.push( ...tasksArray );
+}
 
-	config.queue.unshift( callback );
-	priorityCount++;
+/**
+ * Return the number of tasks remaining in the task queue to be processed.
+ * @return {Number}
+ */
+function taskQueueLength() {
+	return taskQueue.length;
 }
 
 /**
- * Adds a function to the ProcessingQueue for execution.
- * @param {Function|Array} callback
- * @param {Boolean} priority
+ * Adds a tests to the TestQueue for execution.
+ * @param {Array} testTasksArray
```

Comment:
I could be wrong, but isn't this always going to be a function now? We pass the `runTest` function into here, and then later `shift` them off and then pass the expanded array to `addToTaskQueue`, right?

Reply:
Correct! I renamed this to `testTasksFunc`